adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+------------------------ MODULE query_deduplication ------------------------
+
+
+EXTENDS Integers, Sequences, FiniteSets, TLC
+
+CONSTANT Proc
+CONSTANT Keys
+(*
+
+\* from https://probablydance.com/2020/10/31/using-tla-in-the-real-world-to-understand-a-glibc-bug/
+--algorithm Spinlock 
+{
+  variables
+    cache_lock = 0;
+    wait_map_lock = 1;
+    locks = [i \in {cache_lock, wait_map_lock} |-> FALSE];
+    \* set to false for every possible key at first
+    cache = [ i \in Keys |-> FALSE ];
+    wait_map = [ i \in Keys |-> {} ];
+    channels = [i \in Proc |-> FALSE ];
+    tested_keys = {};
+ 
+   define {
+
+    MutualExclusion == \A i, j \in Proc : 
+                         (i # j) => ~ /\ pc[i] = "cs"
+                                      /\ pc[j] = "cs"
+                                      
+    DeadlockFreedom == 
+        \A i \in Proc : 
+          (pc[i] = "exchange") ~> (\E j \in Proc : pc[j] = "cs")
+    
+    
+    \*AllFinished == <>(\A proc \in Proc: pc[proc] = "Done")
+    \*AllKeysLoaded == <>(\A k \in tested_keys: cache[k])
+  };
+
+  procedure lock(idx = 0)
+  {
+    exchange:
+      await ~locks[idx];
+    check:
+      if(~locks[idx]) {
+        locks[idx] := TRUE;
+        return;
+      } else {
+        goto exchange;
+      };
+  };
+
+  procedure unlock(idx = 0)
+  {
+    reset_state:
+      locks[idx] := FALSE;
+      return;
+  }
+  
+  fair process (P \in Proc)
+  variables Key \in Keys, pr = 0;
+  {
+  
+ 
+    start:
+        \* let mut locked_cache = self.cached.lock().await;
+        call lock(cache_lock);
+
+    cache_locked:
+        \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+        if (cache[Key]) {
+            \* if there is a value for the key, we test two possible behaviours
+            \* expire the key, or assume a normal task that will just return after
+            \* receiving the value
+            either {
+                \* expire one key
+                expiration_lock_1:
+                    cache[Key] := FALSE;
+                    call unlock(cache_lock);
+
+                expiration_done:
+                    goto finished;
+            }
+            or {
+                \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+                call unlock(cache_lock);
+                cache_hit:
+                    goto finished;
+            }
+        };
+
+    cache_miss:
+        \* we test at the end if all keys have been loaded
+        tested_keys := tested_keys \union {Key};
```

Comment:
this remains from an assertion I was testing before adding expiration: at the end of the run, the cache should have a value for all the keys that were requested.
Now the model is simpler: we test that the execution terminates (if everything worked correctly every process will reach the end state)